core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,404 @@
+    @Override
+    public String toString() {
+      return getClass().getSimpleName() + '{' +
+          "expireAfterAccessMs=" + expireAfterAccessMs + ", " +
+          "maxContentLength=" + maxContentLength + ", " +
+          "maxTotalBytes=" + maxTotalBytes + ", " +
+          "cacheStats=" + cache.stats() + '}';
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length : wrappedInputFile.getLength();
+    }
+
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        return wrappedInputFile.newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      }
+    }
+
+    @Override
+    public String location() {
+      return wrappedInputFile.location();
+    }
+
+    @Override
+    public boolean exists() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return buf != null || wrappedInputFile.exists();
+    }
+
+    public long getLengthChecked() throws IOException {
+      try {
+        return getLength();
+      } catch (RuntimeIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    private byte[] getFileContent() {
+      try {
+        int totalByteRead = 0;
+        boolean eos = false;
+        int fileLength = (int) getLength();
+        byte[] buf = new byte[fileLength];
+        SeekableInputStream stream = wrappedInputFile.newStream();
+        while (totalByteRead < fileLength && !eos) {
+          int byteRead = stream.read(buf, totalByteRead, fileLength - totalByteRead);
+          if (byteRead < 0) {
+            // Somehow we hit EOS and ended up short. Truncate buf.
+            buf = Arrays.copyOf(buf, totalByteRead);
+            eos = true;
+          } else {
+            totalByteRead += byteRead;
+          }
+        }
+
+        stream.close();
+        return buf;
+      } catch (IOException ex) {
+        throw new RuntimeIOException(ex);
+      }
+    }
+
+    private SeekableInputStream cachedStream() throws IOException {
+      try {
+        byte[] buf = contentCache.cache().get(location(), k -> getFileContent());
+        LOG.debug("Cache stats: {}", contentCache.cache().stats());
+        return new SeekableCachedInputStream(new SeekableByteArrayInputStream(buf));
```

Comment:
I think this should use a generic in-memory input stream, rather than one that is specific to caching.

Reply:
Refactored to use `ByteBufferInputStream`.